src/GafferOSLUI/OSLObjectUI.cpp

```diff
@@ -0,0 +1,271 @@
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+
+#include "GafferOSL/OSLObject.h"
+
+#include "GafferUI/Nodule.h"
+#include "GafferUI/NoduleLayout.h"
+#include "GafferUI/PlugAdder.h"
+
+#include "Gaffer/Metadata.h"
+#include "Gaffer/MetadataAlgo.h"
+#include "Gaffer/ScriptNode.h"
+#include "Gaffer/UndoScope.h"
+#include "Gaffer/NameValuePlug.h"
+#include "Gaffer/PlugAlgo.h"
+
+#include "IECore/CompoundData.h"
+
+#include "boost/bind.hpp"
+
+using namespace std;
+using namespace Gaffer;
+using namespace GafferUI;
+
+namespace
+{
+
+class OSLObjectPlugAdder : public PlugAdder
+{
+
+	public :
+
+		OSLObjectPlugAdder( GraphComponentPtr plugsParent )
+			:   m_plugsParent( IECore::runTimeCast<Plug>( plugsParent ) )
+		{
+			if( ! m_plugsParent )
+			{
+				throw IECore::Exception( "OSLObjectUI::PlugAdder constructor must be passed plug" );
+			}
+			buttonReleaseSignal().connect( boost::bind( &OSLObjectPlugAdder::buttonRelease, this, ::_2 ) );
+		}
+
+	protected :
+
+		bool canCreateConnection( const Plug *endpoint ) const override
+		{
+			vector<PlugPtr> plugs = proxyPlugs( endpoint );
+			return !plugs.empty();
+		}
+
+		void createConnection( Plug *endpoint ) override
+		{
+			vector<PlugPtr> proxies = proxyPlugs( endpoint );
+			vector<std::string> names;
+			for( auto &i : proxies )
+			{
+				names.push_back( i->getName() );
+			}
+
+			std::string picked = nameMenuSignal()( "Connect To", names );
```

Comment:
I'm confused. I thought we added `nameMenuSignal()` so you didn't need to build the proxy plugs, but now we're building the proxy plugs, then building a list of strings from them, then using the new menu, and then searching back in the proxies for the right name?

Reply:
Gah, you're right.  I had originally thought that I would just prep a list of matching options as Data, and just do the createPlugFromData after picking one an option from the menu.  But I can't conveniently represent a Closure as Data, so it ended up keeping the proxyPlugs method I had orginally made to work around the lack of nameMenuSignal.  I guess the quickest solution is just to delete nameMenuSignal?

Reply:
> I guess the quickest solution is just to delete nameMenuSignal?

Yeah, let's do that.

Reply:
Conclusion from conversation with John:  the new signal is useful, whether we use it here or not.  I'll rename it from nameMenuSignal to just menuSignal, and include the commit, but reassess whether I actually use it in OSLObjectUI